File: motor_metrics/derive.py
"""Resampling, smoothing and differentiation for landmark trajectories.

Sway velocity and movement smoothness both need a *derivative* of a landmark path, and
landmark paths are noisy: differentiating them raw amplifies the noise more than the
signal. So everything routes through one filter chain — resample to a uniform grid,
then Savitzky-Golay.

**Why the uniform grid.** Camera frames land at roughly :data:`FS` but not exactly:
``Recording.fps()`` is instantaneous ``1000/diff(timestamps_ms)`` and wobbles. A
Savitzky-Golay derivative takes a single scalar ``delta`` (one sample spacing) and
SPARC's FFT assumes uniform spacing outright; both quietly produce wrong numbers on a
jittery timebase rather than complaining.

The grid is 15 Hz because that is what the capture hardware actually sustains — the phone
measured 15-20 fps and the desktop CLIs now request 15 to match. Resampling *up* to a
faster grid than the camera delivers does not add information: it manufactures samples
that are exact linear combinations of their neighbours, and the filter below then averages
over samples that are partly each other, at which point the gain table stops describing
the chain.

**Why the constants are constants.** :data:`FS`, :data:`WINDOW_S` and :data:`POLY` are
module-level and not meant to be passed per call. A SPARC score or a sway velocity is
only comparable against another one computed through an identical filter chain, and
these numbers get compared across months of sessions. Exposing them as tweakable
per-call defaults invites two baselines that differ because of a filter setting rather
than because of Remy. If a constant must change, change it once, here — and treat every
number computed before the change as belonging to a different scale.

**:data:`FS` and :data:`WINDOW_S` are a pair.** The window is specified in seconds but
applied in samples, so halving the rate halves the sample count: at ``WINDOW_S = 0.25`` a
15 Hz grid gives ``int(0.25 * 15) = 3`` samples, and three samples fit a quadratic
*exactly*. The filter would become the identity and stop filtering, without raising, and
every derivative would silently become a raw central difference. Only two non-degenerate
windows are reachable at 15 Hz — 5 samples (0.333 s) and 7 (0.467 s) — and 0.35 selects
the 5, deliberately sitting mid-bracket rather than at the 0.3333 boundary where one
float rounding would collapse it. See ``WindowLengthTests``, which pins both directions.

**What the chain costs.** At the shipped constants (15 Hz, 0.333 s window, quadratic)
the first-derivative gain rolls off with movement frequency — measured against an
analytic sine, and pinned by ``test_derivative_gain_rolls_off_with_frequency``::

    0.25 Hz  0.994     1.0 Hz  0.904     2.0 Hz  0.652
    0.50 Hz  0.975     1.5 Hz  0.793     3.0 Hz  0.339

So postural sway (well under 1 Hz) is measured essentially unattenuated, while the fast
content of a transition is damped by tens of percent. This is a *bias*, not noise: it is
identical for every trial through the same chain, so it cancels in the within-child
comparisons this package is for, and it is one more reason those constants must not
drift between sessions. It also means these magnitudes are not comparable to figures
produced by any other filter chain, published or otherwise.

The window spans 0.333 s rather than the 0.233 s it spanned at 30 Hz, so this table rolls
off sooner than its predecessor: what was 0.95 at 1 Hz is 0.90, and 3 Hz is damped to a
third rather than to two thirds. Sway is unaffected in practice; fast transition content
is measured more conservatively than it was. The alternative 7-sample window was rejected
on measurement — it reproduces the old table compressed *exactly* 2x in frequency, which
would have halved the package's usable bandwidth to buy noise rejection this chain does
not need.

Nothing here interpolates across a dropout. A gap in tracking is a gap; bridging it
would invent the movement that happened inside it. Callers pick a contiguous good run
first (:func:`motor_metrics.quality.longest_run`) and resample within it.
"""

import numpy as np
from scipy.signal import savgol_filter

FS = 15.0  # Hz, the uniform grid every derived quantity is computed on
WINDOW_S = 0.35  # Savitzky-Golay window, in seconds
POLY = 2  # Savitzky-Golay polynomial order


def window_length(fs: float = FS, window_s: float = WINDOW_S, poly: int = POLY) -> int:
    """Savitzky-Golay window in samples: odd, and greater than ``poly``, as scipy requires.

    **The degenerate case is real and silent.** ``poly + 1`` samples fit a polynomial of
    order ``poly`` *exactly*, so a window that small is an interpolation, not a fit:
    ``deriv=0`` returns the input untouched and ``deriv=1`` collapses to a plain central
    difference. Nothing raises — the chain simply stops filtering. At ``POLY=2`` that is
    ``w == 3``, which ``window_s=0.25`` produces at any ``fs`` in [12, 20), i.e. exactly the
    range a 15 Hz grid sits in. :data:`WINDOW_S` is therefore paired with :data:`FS` and the
    two must be changed together; the pairing is pinned by ``WindowLengthTests``.
    """
    w = max(int(window_s * fs), poly + 1)
    return w + 1 if w % 2 == 0 else w


def resample_uniform(t_ms, x, fs: float = FS):
    """Linearly resample ``x`` from timestamps ``t_ms`` onto a uniform ``fs`` grid.

    ``x`` is ``(N,)`` or ``(N, K)``; returns ``(t_s, x_uniform)`` with ``t_s`` in
    seconds from the first sample. Returns empty arrays when fewer than two samples are
    given — there is nothing to interpolate between, and a caller asking for a
    derivative of one point should get an empty answer, not an exception.

    NaN inputs propagate to their neighbouring intervals rather than being dropped:
    silently interpolating across missing frames would fabricate the very movement the
    tracker failed to see.
    """
    t = np.asarray(t_ms, dtype=np.float64)
    x = np.asarray(x, dtype=np.float64)
    single = x.ndim == 1
    cols = x.reshape(-1, 1) if single else x

    if t.size < 2:
        return np.empty(0), np.empty(0) if single else np.empty((0, cols.shape[1]))

    t_s = (t - t[0]) / 1000.0
    span = float(t_s[-1])
    count = int(np.floor(span * fs)) + 1
    if count < 2:
        return np.empty(0), np.empty(0) if single else np.empty((0, cols.shape[1]))

    grid = np.arange(count) / fs
    out = np.stack([np.interp(grid, t_s, col) for col in cols.T], axis=1)
    return grid, (out[:, 0] if single else out)


def smooth(x, fs: float = FS, *, window_s: float = WINDOW_S, poly: int = POLY, deriv: int = 0):
    """Savitzky-Golay filter (or its ``deriv``-th derivative) along axis 0.

    Assumes ``x`` is already on a uniform ``fs`` grid — pass it through
    :func:`resample_uniform` first. With ``deriv=1`` the result is in units per second.

    Returns an all-NaN array of the input's shape when the input is shorter than the
    window, rather than raising as ``savgol_filter`` does. Short segments are ordinary
    here — a two-second sit is a real trial, and a two-frame one is a mis-marked
    annotation. Neither should take down a report table that has 40 other rows in it.
    """
    x = np.asarray(x, dtype=np.float64)
    w = window_length(fs, window_s, poly)
    if x.shape[0] < w:
        return np.full(x.shape, np.nan)
    return savgol_filter(x, w, poly, deriv=deriv, delta=1.0 / fs, axis=0)


def velocity(t_ms, p, fs: float = FS):
    """Velocity of a path: ``(t_s, components, speed)``.

    ``p`` is ``(N, K)`` positions at ``t_ms``. Returns the uniform time grid, the
    ``(M, K)`` per-axis velocity, and the ``(M,)`` scalar speed (the norm), all in
    units per second. Speed is NaN throughout when the segment is shorter than the
    smoothing window.
    """
    t_s, uniform = resample_uniform(t_ms, p, fs=fs)
    if t_s.size == 0:
        k = np.asarray(p).reshape(len(np.asarray(p)), -1).shape[1]
        return np.empty(0), np.empty((0, k)), np.empty(0)
    v = smooth(uniform, fs=fs, deriv=1)
    return t_s, v, np.linalg.norm(v, axis=1)

File: android/metrics/src/main/kotlin/dev/remapy/metrics/Derive.kt
package dev.remapy.metrics

import kotlin.math.abs
import kotlin.math.floor
import kotlin.math.max

/**
 * Resampling, smoothing and differentiation for landmark trajectories.
 *
 * Port of `motor_metrics/derive.py`. Read that module's docstring first — it carries the
 * reasoning, the measured derivative-gain table, and the reason [FS], [WINDOW_S] and [POLY]
 * are constants rather than parameters. The short version: two numbers are comparable only
 * through an identical filter chain, and these get compared across months of sessions.
 *
 * **This file is the one with real risk in it.** Everything else in the kernel is arithmetic
 * that either matches or obviously does not. Savitzky-Golay has an *edge* convention, and
 * `LIVE_LAG` — the whole reason a live readout is trustworthy — is defined by it. scipy's
 * default `mode="interp"` does not pad the signal: it fits one polynomial of order [POLY] to
 * the first `windowLength` samples and evaluates it at each of the first `halfWindow`
 * positions, and likewise at the tail. A port that mirrors, reflects or zero-pads instead
 * will agree on every interior sample and disagree on exactly the `windowLength / 2` that the
 * live instantaneous readout is taken from. `SavgolTest` and `LiveLagIdentityTest` pin it.
 */
object Derive {
    /**
     * Hz, the uniform grid every derived quantity is computed on.
     *
     * 15 because that is what the capture hardware sustains — the phone measured 15-20 fps
     * and both it and the desktop CLIs now request 15. Resampling *up* to a grid faster than
     * the camera delivers adds no information; it manufactures samples that are exact linear
     * combinations of their neighbours, and the filter then averages over samples that are
     * partly each other.
     */
    const val FS: Double = 15.0

    /**
     * Savitzky-Golay window, in seconds. **Paired with [FS]** — see [windowLength].
     */
    const val WINDOW_S: Double = 0.35

    /** Savitzky-Golay polynomial order. */
    const val POLY: Int = 2

    /**
     * Savitzky-Golay window in samples: odd, and greater than [POLY], as scipy requires.
     *
     * `max(int(window_s * fs), poly + 1)` in Python truncates toward zero; [windowLength]
     * only ever sees positive input, so [floor] matches.
     *
     * **The degenerate case is real and silent.** `poly + 1` samples fit a polynomial of order
     * [POLY] *exactly*, so a window that small is an interpolation rather than a fit: `deriv=0`
     * returns its input untouched and `deriv=1` collapses to a plain central difference, with
     * nothing raised and nothing NaN. At [POLY] = 2 that is `w == 3`, which `windowS = 0.25`
     * produces at any `fs` in [12, 20) — exactly where a 15 Hz grid sits. [WINDOW_S] is
     * therefore paired with [FS] and the two move together; `ConstantsTest` pins the pair
     * against the Python side's own value.
     */
    fun windowLength(fs: Double = FS, windowS: Double = WINDOW_S, poly: Int = POLY): Int {
        val w = max(floor(windowS * fs).toInt(), poly + 1)
        return if (w % 2 == 0) w + 1 else w
    }

    /**
     * Linearly resample [x] from timestamps [tMs] onto a uniform [fs] grid.
     *
     * Returns the grid in seconds from the first sample, and the resampled columns. Returns
     * empty arrays when fewer than two samples are given, or when the span is shorter than
     * one grid step — a caller asking for the derivative of a single point should get an
     * empty answer, not an exception.
     *
     * NaN inputs propagate to their neighbouring intervals rather than being dropped.
     * Silently interpolating across missing frames would fabricate the very movement the
     * tracker failed to see, which is the same refusal `longestRun` makes.
     */
    fun resampleUniform(tMs: DoubleArray, x: Matrix, fs: Double = FS): Pair<DoubleArray, Matrix> {
        val cols = x.cols
        if (tMs.size < 2 || x.rows < 2) return DoubleArray(0) to Matrix(0, cols)

        val tS = DoubleArray(tMs.size) { (tMs[it] - tMs[0]) / 1000.0 }
        val span = tS[tS.size - 1]
        val count = floor(span * fs).toInt() + 1
        if (count < 2) return DoubleArray(0) to Matrix(0, cols)

        val grid = DoubleArray(count) { it / fs }
        val out = Matrix(count, cols)
        for (c in 0 until cols) {
            for (i in 0 until count) out[i, c] = interp(grid[i], tS, x, c)
        }
        return grid to out
    }

    /** [resampleUniform] for a single column. */
    fun resampleUniform(tMs: DoubleArray, x: DoubleArray, fs: Double = FS): Pair<DoubleArray, DoubleArray> {
        val (t, m) = resampleUniform(tMs, Matrix.ofColumn(x), fs)
        return t to m.column(0)
    }

    /**
     * `np.interp` for one query point: clamped at both ends, NaN-propagating.
     *
     * numpy clamps outside the sample range to the endpoint values rather than
     * extrapolating. The grid here never exceeds the range by construction, but matching
     * the semantics costs nothing and removes a way for the two implementations to differ
     * only on a rounding edge.
     */
    private fun interp(at: Double, xs: DoubleArray, values: Matrix, col: Int): Double {
        val n = xs.size
        if (at <= xs[0]) return values[0, col]
        if (at >= xs[n - 1]) return values[n - 1, col]

        var lo = 0
        var hi = n - 1
        while (hi - lo > 1) {
            val mid = (lo + hi) ushr 1
            if (xs[mid] <= at) lo = mid else hi = mid
        }
        val x0 = xs[lo]
        val x1 = xs[hi]
        val y0 = values[lo, col]
        val y1 = values[hi, col]
        if (x1 == x0) return y0
        return y0 + (y1 - y0) * (at - x0) / (x1 - x0)
    }

    /**
     * Savitzky-Golay filter (or its [deriv]-th derivative) along axis 0.
     *
     * Assumes [x] is already on a uniform [fs] grid — run it through [resampleUniform]
     * first. With `deriv = 1` the result is in units per second.
     *
     * Returns an all-NaN array of the input's shape when the input is shorter than the
     * window, rather than throwing as scipy's `savgol_filter` does. Short segments are
     * ordinary: a two-second sit is a real trial and a two-frame one is a mis-marked
     * annotation, and neither should take down a report that has 40 other rows in it.
     */
    fun smooth(
        x: Matrix,
        fs: Double = FS,
        windowS: Double = WINDOW_S,
        poly: Int = POLY,
        deriv: Int = 0,
    ): Matrix {
        val w = windowLength(fs, windowS, poly)
        if (x.rows < w) return Matrix.filled(x.rows, x.cols, Double.NaN)

        val delta = 1.0 / fs
        val half = w / 2
        val interior = savgolCoefficients(w, poly, deriv, delta, half.toDouble())
        val out = Matrix(x.rows, x.cols)

        for (c in 0 until x.cols) {
            val col = x.column(c)
            // Interior: a plain FIR convolution with the centred coefficients.
            for (i in half until x.rows - half) {
                var acc = 0.0
                for (k in 0 until w) acc += interior[k] * col[i - half + k]
                out[i, c] = acc
            }
            // Edges: scipy's mode="interp" fits ONE polynomial to the first (last) `w`
            // samples and evaluates it at each edge position, rather than extending the
            // signal. Equivalent to a per-position coefficient vector over that same block.
            for (i in 0 until half) {
                val coef = savgolCoefficients(w, poly, deriv, delta, i.toDouble())
                var acc = 0.0
                for (k in 0 until w) acc += coef[k] * col[k]
                out[i, c] = acc
            }
            for (i in 0 until half) {
                val pos = w - 1 - i
                val coef = savgolCoefficients(w, poly, deriv, delta, pos.toDouble())
                var acc = 0.0
                for (k in 0 until w) acc += coef[k] * col[x.rows - w + k]
                out[x.rows - 1 - i, c] = acc
            }
        }
        return out
    }

    /** [smooth] for a single column. */
    fun smooth(
        x: DoubleArray,
        fs: Double = FS,
        windowS: Double = WINDOW_S,
        poly: Int = POLY,
        deriv: Int = 0,
    ): DoubleArray = smooth(Matrix.ofColumn(x), fs, windowS, poly, deriv).column(0)

    /**
     * Savitzky-Golay coefficients: the [deriv]-th derivative of the least-squares
     * polynomial fit over a [windowLength]-sample block, evaluated at sample [pos].
     *
     * `pos = windowLength / 2` gives scipy's centred `savgol_coeffs`; other values give the
     * edge coefficients `mode="interp"` uses. Solved as a small normal-equation system —
     * the window is 5 samples at the shipped constants, so a general least-squares routine
     * would be more machinery than the problem has.
     */
    fun savgolCoefficients(
        windowLength: Int,
        poly: Int,
        deriv: Int,
        delta: Double,
        pos: Double,
    ): DoubleArray {
        require(windowLength > poly) { "windowLength must exceed poly, got $windowLength and $poly." }
        if (deriv > poly) return DoubleArray(windowLength)

        val order = poly + 1
        // Vandermonde on the sample offsets: A[i][j] = (i - pos)^j.
        val a = Array(windowLength) { i ->
            DoubleArray(order) { j -> powInt(i - pos, j) }
        }
        // Normal equations (A^T A) c = A^T e_deriv, where the right-hand side selects the
        // `deriv`-th Taylor coefficient of the fitted polynomial.
        val ata = Array(order) { r -> DoubleArray(order) { c ->
            var acc = 0.0
            for (i in 0 until windowLength) acc += a[i][r] * a[i][c]
            acc
        } }
        val rhs = DoubleArray(order)
        rhs[deriv] = factorial(deriv) / powInt(delta, deriv)

        val alpha = solveSymmetric(ata, rhs)
        // c_k = sum_j alpha_j * A[k][j]: the fitted derivative as a linear functional of the
        // samples, which is exactly the FIR kernel.
        return DoubleArray(windowLength) { k ->
            var acc = 0.0
            for (j in 0 until order) acc += alpha[j] * a[k][j]
            acc
        }
    }

    /** Gaussian elimination with partial pivoting. Systems here are (poly + 1) square. */
    private fun solveSymmetric(matrix: Array<DoubleArray>, rhs: DoubleArray): DoubleArray {
        val n = rhs.size
        val m = Array(n) { r -> DoubleArray(n + 1) { c -> if (c < n) matrix[r][c] else rhs[r] } }
        for (col in 0 until n) {
            var pivot = col
            for (r in col + 1 until n) if (abs(m[r][col]) > abs(m[pivot][col])) pivot = r
            val tmp = m[col]; m[col] = m[pivot]; m[pivot] = tmp
            val p = m[col][col]
            if (p == 0.0) continue
            for (r in 0 until n) {
                if (r == col) continue
                val factor = m[r][col] / p
                if (factor == 0.0) continue
                for (c in col..n) m[r][c] -= factor * m[col][c]
            }
        }
        return DoubleArray(n) { if (m[it][it] == 0.0) 0.0 else m[it][n] / m[it][it] }
    }

    private fun powInt(base: Double, exponent: Int): Double {
        var acc = 1.0
        repeat(exponent) { acc *= base }
        return acc
    }

    private fun factorial(n: Int): Double {
        var acc = 1.0
        for (i in 2..n) acc *= i
        return acc
    }

    /**
     * Velocity of a path: the uniform time grid, the per-axis velocity, and the scalar speed.
     *
     * All in units per second. Speed is NaN throughout when the segment is shorter than the
     * smoothing window.
     */
    fun velocity(tMs: DoubleArray, p: Matrix, fs: Double = FS): Triple<DoubleArray, Matrix, DoubleArray> {
        val (tS, uniform) = resampleUniform(tMs, p, fs)
        if (tS.isEmpty()) return Triple(DoubleArray(0), Matrix(0, p.cols), DoubleArray(0))
        val v = smooth(uniform, fs, deriv = 1)
        val speed = DoubleArray(v.rows) { i ->
            var acc = 0.0
            for (c in 0 until v.cols) acc += v[i, c] * v[i, c]
            kotlin.math.sqrt(acc)
        }
        return Triple(tS, v, speed)
    }
}
